python/lsst/dax/ppdb/sql/_chunk_exporter.py

```diff
@@ -0,0 +1,187 @@
+                        "schema_version": str(self.schema_version),
+                        "status": "pending",
+                        "table": table_name,
+                    }
+                    self._upload_table_to_gcs(arrow_table, gcs_path, metadata)
+                    gcs_paths[gcs_path] = metadata
+                else:
+                    _LOG.warning("GCS operations are disabled. Skipping upload.")
+
+            self._set_ready(gcs_paths)
+        except Exception as e:
+            _LOG.error("Failed to store replica chunk: %s", e)
+            # Rollback the GCS upload for the chunk if there was an error.
+            if gcs_paths:
+                _LOG.info("Attempting to delete GCS paths: %s", gcs_paths)
+                self._delete_objects(gcs_paths)
+            raise
+
+        # Update the database to indicate that the chunk has been exported.
+        with self._engine.begin() as connection:
+            self._store_insert_id(replica_chunk, connection, update)
+
+    @classmethod
+    def _convert_to_arrow(cls, table_data: ApdbTableData) -> pa.Table:
+        rows = list(table_data.rows())
+        column_names = list(table_data.column_names())
+        _LOG.info("Converting %d rows with %d columns to Arrow Table", len(rows), len(column_names))
+        columns = list(zip(*rows))
+        arrays = [pa.array(col) for col in columns]
+        return pa.table(dict(zip(column_names, arrays)))
+
+    def _delete_objects(self, gcs_paths: list[str]) -> None:
+        if not storage:
+            _LOG.warning("GCS operations are disabled. Cannot delete objects.")
+            return
+
+        for gcs_path in gcs_paths:
+            blob = self.bucket.blob(gcs_path)
+            blob.delete()
+            _LOG.info("Deleted GCS path %s", gcs_path)
+
+    def _set_ready(self, gcs_paths: dict[str, dict[str, str]]) -> None:
+        if not storage:
+            _LOG.warning("GCS operations are disabled. Cannot set ready.")
+            return
+
+        for gcs_path, metadata in gcs_paths.items():
+            blob = self.bucket.blob(gcs_path)
+            blob.metadata = metadata
+            blob.metadata["status"] = "ready"
+            try:
+                blob.patch()
+                _LOG.info("Set GCS path %s to ready", gcs_path)
+                _LOG.info("New metadata: %s", blob.metadata)
+            except Exception as e:
+                _LOG.error("Failed to set GCS path %s to ready: %s", gcs_path, e)
+                raise
+
+    def _make_gcs_path(self, table_name: str, chunk_id: int) -> str:
+        today = datetime.today()
+        year = today.strftime("%Y")
+        month = today.strftime("%m")
+        day = today.strftime("%d")
+        return f"{self.folder_name}/{year}/{month}/{day}/{str(chunk_id)}/{table_name}.parquet"
+
+    def _upload_table_to_gcs(self, table: pa.Table, gcs_path: str, metadata: dict[str, str]) -> None:
+        with BytesIO() as buf:
```

Comment:
Again, you want to write files to local storage before sending them to S3. This will save APDB-side traffic in case S3 has issues and will make things more reliable in general.

Reply:
Okay, I will split this up into two separate classes then, one which writes the files locally, and another which uploads them to GCS.